linker/shared/src/main/scala/org/scalajs/linker/checker/IRChecker.scala

```diff
@@ -777,18 +649,62 @@ private final class IRChecker(unit: LinkingUnit, logger: Logger) {
     }
 
     tree match {
+      // Definitions
+
+      case VarDef(ident, _, vtpe, _, rhs) =>
+        checkDeclareLocalVar(ident)
+        typecheckExpect(rhs, env, vtpe)
+
       // Control flow constructs
 
-      case Block(statsAndExpr) =>
-        val stats :+ expr = statsAndExpr
-        val envAfterStats = stats.foldLeft(env) { (prevEnv, stat) =>
-          typecheckStat(stat, prevEnv)
-        }
-        typecheckExpr(expr, envAfterStats)
+      case Skip() =>
+
+      case Block(trees) =>
+        if (trees.isEmpty)
+          reportError("illegal empty block")
+        typecheckBlockTrees(trees, env)
 
       case Labeled(label, tpe, body) =>
         checkDeclareLabel(label)
-        typecheckExpect(body, env.withLabeledReturnType(label.name, tpe), tpe)
+
+        val expectedType =
+          if (tpe == NoType) AnyType
+          else tpe
```

Comment:
This is a bit suspicious. I guess the deal here is that `return` must take an expression. I wonder if we should require `UndefType` instead (in a different PR, that would be a non-backward compatible change).

Reply:
I agree that it's weird that a `Return` node demands an expression, even when the result of the `Labeled` block is `NoType`.

For what it's worth, in the theory it was the `Return` node that had a specific exception, making `void` (`NoType`) invalid:
![image](https://user-images.githubusercontent.com/535934/109138707-f0769f80-775a-11eb-92e7-e4d51f203db9.png)
(note the `S ≠ void`)

Perhaps we could align the code of the IR checker with the typing rules a bit better by moving this change to the checking of `Return`.

If we have to change something in the future, I would rather relax the restriction on `Return` nodes, and make them accept "expressions" of type `NoType` if the type of the `Labeled` is `NoType`. That would be much nicer than requiring `UndefType` I believe. Plus, it would be backward compatible.